sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -0,0 +1,277 @@
+#include "sw/device/silicon_creator/lib/shutdown.h"
+
+#include <assert.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/alert.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+
+#include "alert_handler_regs.h"
+#include "flash_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"
+#include "otp_ctrl_regs.h"
+#include "sram_ctrl_regs.h"
+
+
+static_assert(
+  ALERT_HANDLER_ALERT_CLASS_MULTIREG_COUNT <= OTP_CTRL_PARAM_ROM_ALERT_CLASSIFICATION_SIZE / 4,
+  "More alerts than alert classification OTP words!");
+
+
+#ifdef MOCK_ABS_MMIO
+// If we're building as a unit test, rename the shutdown functions so they
+// can be mocked and/or tested individually.
+#define SHUTDOWN_FUNC(name_) \
+  extern void name_;         \
+  void internal_##name_
+#else
+#define SHUTDOWN_FUNC(name_) static inline void name_
+#endif
+
+// TODO: use the real OTP driver after it's converted to abs_mmio.
+#ifdef MOCK_ABS_MMIO
+extern uint32_t otp_read32(uint32_t address);
+#else
+inline uint32_t otp_read32(uint32_t address) {
+  return abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR +
+                         OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
+}
+#endif
+
+// Convert the alert class to an index.
+// This is required because I expect to change the constant definitions in
+// alert_class_t to have reasonable hamming distances.
+static size_t clsindex(alert_class_t cls) {
+  switch (cls) {
+    case kAlertClassA:
+      return 0;
+    case kAlertClassB:
+      return 1;
+    case kAlertClassC:
+      return 2;
+    case kAlertClassD:
+      return 3;
+    default:
+      return 0;
+  }
+}
+
+rom_error_t shutdown_init(lifecycle_state_t lc_state) {
+  const alert_class_t kClasses[] = {
+      kAlertClassA,
+      kAlertClassB,
+      kAlertClassC,
+      kAlertClassD,
+  };
+  uint32_t shift = 0;
+  uint32_t class_enable, class_escalate;
```

Comment:
Please don't pre-declare variables; definite them where they are first used, if possible (and move declarations as close as possible to first use otherwise).

Reply:
Fixed.